src/expr/nary_term_util.cpp

```diff
@@ -0,0 +1,249 @@
+        return false;
+      }
+      for (const Node& cn : cur)
+      {
+        if (isListVar(cn))
+        {
+          itc = context.find(cn);
+          if (itc == context.end())
+          {
+            context[cn] = cur.getKind();
+          }
+          else if (itc->second != cur.getKind())
+          {
+            return false;
+          }
+          continue;
+        }
+        visit.push_back(cn);
+      }
+    }
+  } while (!visit.empty());
+  return true;
+}
+
+Node getNullTerminator(Kind k, TypeNode tn)
+{
+  NodeManager* nm = NodeManager::currentNM();
+  Node nullTerm;
+  switch (k)
+  {
+    case OR: nullTerm = nm->mkConst(false); break;
+    case AND:
+    case SEP_STAR: nullTerm = nm->mkConst(true); break;
+    case PLUS: nullTerm = nm->mkConst(Rational(0)); break;
+    case MULT:
+    case NONLINEAR_MULT: nullTerm = nm->mkConst(Rational(1)); break;
+    case STRING_CONCAT:
+      // handles strings and sequences
+      nullTerm = theory::strings::Word::mkEmptyWord(tn);
+      break;
+    case REGEXP_CONCAT:
+      // the language containing only the empty string
+      nullTerm = nm->mkNode(STRING_TO_REGEXP, nm->mkConst(String("")));
+      break;
+    case BITVECTOR_AND:
+      nullTerm = theory::bv::utils::mkOnes(tn.getBitVectorSize());
+      break;
+    case BITVECTOR_OR:
+    case BITVECTOR_ADD:
+    case BITVECTOR_XOR:
+      nullTerm = theory::bv::utils::mkZero(tn.getBitVectorSize());
+      break;
+    case BITVECTOR_MULT:
+      nullTerm = theory::bv::utils::mkOne(tn.getBitVectorSize());
+      break;
+    default:
+      // not handled as null-terminated
+      break;
+  }
+  return nullTerm;
+}
+
+Node narySubstitute(Node src,
+                    const std::vector<Node>& vars,
+                    const std::vector<Node>& subs)
+{
+  // assumes all variables are list variables
+  NodeManager* nm = NodeManager::currentNM();
+  std::unordered_map<TNode, Node> visited;
+  std::unordered_map<TNode, Node>::iterator it;
+  std::vector<TNode> visit;
+  std::vector<Node>::const_iterator itv;
+  TNode cur;
+  visit.push_back(src);
+  do
+  {
+    cur = visit.back();
+    it = visited.find(cur);
+    if (it == visited.end())
+    {
+      // if it is a non-list variable, do the replacement
+      itv = std::find(vars.begin(), vars.end(), cur);
+      if (itv != vars.end())
+      {
+        size_t d = std::distance(vars.begin(), itv);
+        if (!isListVar(vars[d]))
+        {
+          visited[cur] = subs[d];
+          continue;
+        }
+      }
+      visited[cur] = Node::null();
+      visit.insert(visit.end(), cur.begin(), cur.end());
+      continue;
+    }
+    visit.pop_back();
+    if (it->second.isNull())
+    {
+      Node ret = cur;
+      bool childChanged = false;
+      std::vector<Node> children;
+      for (const Node& cn : cur)
+      {
+        // if it is a list variable, insert the corresponding list as children;
+        itv = std::find(vars.begin(), vars.end(), cn);
```

Comment:
To get better performance, we could require that `vars` are `BOUND_VARIABLE`s and then only do this check if `cn` is a bound variable.

Reply:
This is true, although at the moment, I don't want to have this restriction (in risk of forgetting it).